gcc/rust/expand/rust-macro-builtins-asm.cc

```diff
@@ -16,5 +16,528 @@
+      //  {
+      //           span: p.token.span,
+      //       }));
+    }
+  if (!parser.skip_token (RIGHT_PAREN))
+    {
+      // TODO: we expect a left parenthesis here, please return the correct
+      // error.
+      return tl::nullopt;
+    }
+
+  return regClass;
+}
+
+int
+parse_operand (Parser<MacroInvocLexer> &parser, TokenId last_token_id,
+	       InlineAsmContext &inlineAsmCtx)
+{
+  return 0;
+}
+
+// From rustc
+tl::optional<AST::InlineAsmOperand>
+parse_reg_operand (Parser<MacroInvocLexer> &parser, TokenId last_token_id,
+		   InlineAsmContext &inlineAsmCtx)
+{
+  // let name = if p.token.is_ident() && p.look_ahead(1, |t| *t == token::Eq) {
+  //           let (ident, _) = p.token.ident().unwrap();
+  //           p.bump();
+  //           p.expect(&token::Eq)?;
+  //           allow_templates = false;
+  //           Some(ident.name)
+  //       } else {
+  //           None
+  //       };
+
+  using RegisterType = AST::InlineAsmOperand::RegisterType;
+  AST::InlineAsmOperand reg_operand;
+  auto token = parser.peek_current_token ();
+  auto iden_token = parser.peek_current_token ();
+  auto &inlineAsm = inlineAsmCtx.inlineAsm;
+  if (check_identifier (parser, ""))
+    {
+      auto equal_token = parser.peek_current_token ();
+      if (!parser.skip_token (EQUAL))
+	{
+	  // TODO: Please handle this.
+	  rust_unreachable ();
+	}
+    }
+
+  token = parser.peek_current_token ();
+
+  bool is_explicit_reg = false;
+  bool is_global_asm = inlineAsm.is_global_asm;
+  if (!is_global_asm && check_identifier (parser, "in"))
+    {
+      return tl::nullopt;
+    }
+  else if (!is_global_asm && check_identifier (parser, "out"))
+    {
+      return tl::nullopt;
+    }
+  else if (!is_global_asm && check_identifier (parser, "lateout"))
+    {
+      return tl::nullopt;
+    }
+  else if (!is_global_asm && check_identifier (parser, "inout"))
+    {
+      return tl::nullopt;
+    }
+  else if (!is_global_asm && check_identifier (parser, "inlateout"))
+    {
+      return tl::nullopt;
+    }
+  else if (parser.peek_current_token ()->get_id () == CONST)
+    {
+      rust_unreachable ();
+      // todo: Please handle const
+      return tl::nullopt;
+    }
+  else if (false && check_identifier (parser, "sym"))
+    {
+      // todo: Please handle sym
+      return tl::nullopt;
+    }
+  else if (false && check_identifier (parser, "label"))
+    {
+      // todo: Please handle label
+      return tl::nullopt;
+    }
+  else
+    {
+      return tl::nullopt;
+    }
+  return reg_operand;
+}
+void
+check_and_set (Parser<MacroInvocLexer> &parser, InlineAsmContext &inlineAsmCtx,
+	       AST::InlineAsmOption option)
+{
+  auto &inlineAsm = inlineAsmCtx.inlineAsm;
+  if (inlineAsm.options.count (option) != 0)
+    {
+      // TODO: report an error of duplication
+      rust_error_at (parser.peek_current_token ()->get_locus (),
+		     "the `%s` option was already provided",
```

Comment:
```suggestion
		     "the %qs option was already provided",
```

use `%qs` for quoted strings so that we have the nice unicode quotes if our terminals support them :D